onnx/ir.h

```diff
@@ -0,0 +1,1048 @@
+  // a big pile of messages.
+  std::vector<AVPtr> values_;
+  using iterator = std::vector<AVPtr>::iterator;
+  iterator find(Symbol name, bool required) {
+    auto it = std::find_if(values_.begin(), values_.end(),[&](const AVPtr & v) {
+      return v->name == name;
+    });
+    ONNX_ASSERT(!required || it != values_.end());
+    return it;
+  }
+  using const_iterator = std::vector<AVPtr>::const_iterator;
+  const_iterator find(Symbol name, bool required) const {
+    auto it = std::find_if(values_.begin(), values_.end(),[&](const AVPtr & v) {
+      return v->name == name;
+    });
+    ONNX_ASSERTM(!required || it != values_.end(),
+        "%s:%u: %s: required undefined attribute '%s'", __FILE__, __LINE__, __func__, name.toString());
+    return it;
+  }
+};
+
+
+
+// Each use is represented by this type, see Node::uses()
+// 'user' is the consumer of the value, offset is the index into
+// 'user's input this where the produces will be found.
+struct Use {
+  Use(Node * user, size_t offset)
+  : user(user), offset(offset) {}
+  Node * user;
+  size_t offset;
+};
+
+static inline bool operator==(const Use & a, const Use & b) {
+  return a.user == b.user && a.offset == b.offset;
+}
+
+
+// the list types are intentionally simple, but we type-def
+// them here so if we need to change them, refactoring will be easier
+using node_list = std::vector<Node*>;
+using value_list = std::vector<Value*>;
+using use_list = std::vector<Use>;
+using NodeKind = Symbol;
+
+
+struct Value {
+  ONNX_DISALLOW_COPY_AND_ASSIGN(Value);
+  Value(Node * node_, size_t offset_);
+
+private:
+  friend struct Node;
+  friend struct Graph;
+  Node * node_;
+  size_t offset_;
+  size_t unique_ = 0;          // unique id
+  size_t stage_ = 0;           // 0-forward, 1-backward, 2-double-backward,...
+  use_list uses_;
+  bool has_name_;
+  std::string unique_name_;
+  onnx::TensorProto_DataType elem_type_;
+  std::vector<Dimension> sizes_;
+
+public:
+  Value* setElemType(onnx::TensorProto_DataType elem_type) {
+    elem_type_ = elem_type;
+    return this;
+  }
+  onnx::TensorProto_DataType elemType() const {
+    return elem_type_;
+  }
+  Value* setSizes(const std::vector<Dimension>& sizes) {
+    sizes_ = sizes;
+    return this;
+  }
+  const std::vector<Dimension>& sizes() const {
+    return sizes_;
+  }
+  size_t unique() const {
+    return unique_;
+  }
+  bool has_name() const {
+    return has_name_;
+  }
+  std::string uniqueName() const {
+    if(unique_name_.size() > 0)
+      return unique_name_;
+    return std::to_string(unique());
+  }
+  Value* setUniqueName(const std::string & name) {
+    has_name_ = true;
+    unique_name_ = name;
+    return this;
+  }
+  Value* setStage(size_t s) {
+    stage_ = s;
+    return this;
+  }
+  size_t stage() const {
+    return stage_;
+  }
+  Node* node() {
+    return node_;
+  }
+  size_t offset() const {
+    return offset_;
+  }
+  const Node * node() const {
+    return node_;
+  }
+  Graph * owningGraph();
```

Comment:
I would agree with this if all the nodes were managed separately, but since every Node is memory managed by its owning Graph, I think having the pointer makes sense; you're never supposed to move a Node from one graph to another, and if you do, having owningGraph makes it easy to complain loudly in an assert.

Reply:
Hmmm, not having these APIs in this level (Value) does not bring troubles. By contrast, giving the capability here make things too flexible, guys who doing optimization could now have 3 places to edit a graph, graph, node, and value. We should restrict this somehow.